curvefs/src/metaserver/s3compact_wq_impl.cpp

```diff
@@ -0,0 +1,418 @@
+                                  std::to_string(curr->compaction) + "_" +
+                                  std::to_string(index) + "_" +
+                                  std::to_string(inodeId);
+            std::string buf;
+            const Aws::String aws_key(objName.c_str(), objName.size());
+            int ret = s3Adapter_->GetObject(aws_key, &buf);
+            if (ret != 0) {
+                return ret;
+            } else {
+                uint64_t blockBegin = index * blockSize + curr->chunkoff;
+                uint64_t blockEnd = (index + 1) * blockSize + curr->chunkoff;
+                if (curr->begin >= blockBegin && curr->end <= blockEnd) {
+                    (*fullChunk) += buf.substr(curr->begin - blockBegin,
+                                               curr->end - curr->begin);
+                } else if (curr->begin >= blockBegin && curr->end > blockEnd) {
+                    (*fullChunk) += buf.substr(curr->begin - blockBegin,
+                                               blockEnd - curr->begin);
+                } else if (curr->begin < blockBegin && curr->end > blockEnd) {
+                    (*fullChunk) += buf.substr(0, blockSize);
+                } else if (curr->begin < blockBegin && curr->end <= blockEnd) {
+                    (*fullChunk) += buf.substr(0, curr->end - blockBegin);
+                }
+            }
+        }
+
+        if (curr->chunkid >= *newChunkid) {
+            (*newChunkid) = curr->chunkid;
+            (*newCompaction) = curr->compaction;
+        }
+        if (next != validList.end() && curr->end < next->begin) {
+            // hole, append 0
+            fullChunk->append(next->begin - curr->end, '\0');
+        }
+        curr = next;
+    }
+    // inc compaction
+    (*newCompaction) += 1;
+    return 0;
+}
+
+virtual MetaStatusCode S3CompactWorkQueueImpl::updateInode(
+    CopysetNode* copysetNode, const PartitionInfo& pinfo, const Inode& inode) {
+    UpdateInodeRequest request;
+    request.set_poolid(pinfo.poolid());
+    request.set_copysetId(pinfo.copysetId());
+    request.set_partitionid(pinfo.partitionid());
+    request.set_fsid(pinfo.fsid());
+    request.set_inodeid(inode.inodeid());
+    request.set_s3chunkinfomap(inode.s3chunkinfomap());
+    UpdateInodeResponse response;
+    S3CompactWorkQueueImpl::updateInodeClosure done;
+    auto updateInodeOp = new UpdateInodeOperator(copysetNode, nullptr, &request,
+                                                 &respnse, &done);
+    updateInodeOp->Propose();
+    done.waitRunned();
+    return response.statusCode;
+}
+
+int S3CompactWorkQueueImpl::writeFullChunk(
+    const std::string& fullChunk, uint64_t inodeId, uint64_t blockSize,
+    uint64_t newChunkid, uint64_t newCompaction,
+    std::vector<std::string>* objsAdded) {
+    uint64_t chunkLen = fullChunk.length();
+    for (uint64_t index = 0; index * blockSize < chunkLen; index += 1) {
+        std::string objName =
```

Comment:
extract this into a single function, because line 217 also do the same thing.

Reply:
fixed.